File: src/main/scala/mmlTests/a201401b/java/gpt4/ProgressiveAcceptorImpl.java
package mmlTests.a201401b.java.gpt4;

import java.util.ArrayList;
import java.util.List;

import mmlTests.a201401b.java.Aggregator;
import mmlTests.a201401b.java.ProgressiveAcceptor;
import mmlTests.a201401b.java.ProgressiveFilter;

public class ProgressiveAcceptorImpl<X> implements ProgressiveAcceptor<X> {

    private List<X> list;
    private Aggregator<X> aggregator;
    private ProgressiveFilter<X> filter;
    private int size;

    public ProgressiveAcceptorImpl() {
        list = new ArrayList<>();
    }

    @Override
    public void setProgressiveFilter(ProgressiveFilter<X> filter) {
        if (filter == null) {
            throw new NullPointerException();
        }
        this.filter = filter;
    }

    @Override
    public void setAggregator(Aggregator<X> aggregator) {
        if (aggregator == null) {
            throw new NullPointerException();
        }
        this.aggregator = aggregator;
    }

    @Override
    public void setSize(int size) {
        if (size < 0) {
            throw new IllegalArgumentException();
        }
        this.size = size;
    }

    @Override
    public boolean accept(int pos, X elem) {
        if (aggregator == null || filter == null) {
            throw new IllegalStateException();
        }
        if (pos < 0 || pos > size || pos > list.size()) {
            return false;
        }
        if (!list.isEmpty() && pos > 0) {
            if (!filter.isNextOK(list.get(pos - 1), elem)) {
                return false;
            }
        }
        if (pos < list.size()) {
            list.subList(pos, list.size()).clear();
        }
        list.add(elem);
        return true;
    }

    @Override
    public X aggregateAll() {
        if (list.isEmpty()) throw new IllegalStateException();
        X result = list.get(0);
        for (int i = 1; i < list.size(); i++) {
            result = aggregator.aggregate(result, list.get(i));
        }
        return result;
    }
}

File: src/main/scala/llmTests/a201401b/scala/gpt4/ProgressiveAcceptorImpl.scala
package llmTests.a201401b.scala.gpt4

import llmTests.a201401b.scala.{Aggregator, ProgressiveAcceptor, ProgressiveFilter}

import scala.collection.mutable

class ProgressiveAcceptorImpl[X] extends ProgressiveAcceptor[X] {

  private var aggregator: Option[Aggregator[X]] = None
  private var progressiveFilter: Option[ProgressiveFilter[X]] = None
  private var size: Option[Int] = None
  private var sequence: mutable.ArrayBuffer[X] = mutable.ArrayBuffer.empty[X]

  override def setProgressiveFilter(filter: ProgressiveFilter[X]): Unit =
    if (filter != null) progressiveFilter = Some(filter)
    else throw new NullPointerException

  override def setAggregator(agg: Aggregator[X]): Unit =
    if (agg != null) aggregator = Some(agg)
    else throw new NullPointerException

  override def setSize(s: Int): Unit =
    if (s >= 0) size = Some(s)
    else throw new IllegalArgumentException

  override def accept(pos: Int, elem: X): Boolean = (progressiveFilter, aggregator, size) match {
    case (Some(filter), Some(_), Some(s)) =>
      if (0 <= pos && pos < s) {
        if (pos >= sequence.length) {
          sequence += elem
          if (pos == 0 || filter.isNextOK(sequence(pos - 1), sequence(pos))) true
          else {
            sequence = sequence.dropRight(1)
            false
          }
        } else {
          val accept = if (pos == 0) true else filter.isNextOK(sequence(pos - 1), elem)
          if (accept) {
            sequence = sequence.take(pos) :+ elem
          }
          accept
        }
      } else false
    case _ => throw new IllegalStateException("ProgressiveAcceptor is not properly initialized")
  }

  override def aggregateAll(): X = aggregator match {
    case Some(agg) =>
      if (sequence.isEmpty) throw new IllegalStateException("Nothing to aggregate")
      else if (sequence.length == 1) sequence.head
      else sequence.reduceLeft(agg.aggregate)
    case None => throw new NullPointerException("Aggregator is missing")
  }
}